pygamelib/gfx/ui.py

```diff
@@ -2637,239 +2568,380 @@ def show(self):  # pragma: no cover
-            self.__orientation = orientation
-        self._config = None
-        if isinstance(config, UiConfig):
-            self._config = config
-        self.__color = core.Color()
-        self.__sprixel = core.Sprixel(" ")
-        self.__red = 128
-        self.__green = 128
-        self.__blue = 128
-        self.__selection = 0
+        if config is None:
+            config = UiConfig.instance()
+        if width is None:
+            width = maximum_width
+        if height is None:
+            height = maximum_height
+        if minimum_width is None:
+            minimum_width = maximum_width
+        if minimum_height is None:
+            minimum_height = maximum_height
+
+        super().__init__(
+            width=width,
+            height=height,
+            minimum_width=minimum_width,
+            minimum_height=minimum_height,
+            maximum_width=maximum_width,
+            maximum_height=maximum_height,
+            bg_color=config.input_bg_color,
+            config=config,
+        )
+        self.__choices = []
+        if choices is not None and type(choices) is list:
+            self.__choices = choices
+        self.__current_choice = 0
+        self.__current_page = 0
+        self.__cache = []
+        self._build_cache()
+        self.__items_per_page = max(
+            int(self.maximum_height / 2 * self.maximum_width / 2), 1
+        )  # set floor of 1 to avoid ZeroDivisionError
+        # config.game.log(f"items per page={self.__items_per_page}")
+
+    def _build_cache(self):
+        self.__cache = []
+        for choice in self.__choices:
+            s = choice
+            if type(s) is str:
+                s = core.Sprixel(choice)
+            # TODO: For v1 we only support sprixels of size=1
+            # Reason is that some emojis are returning a length of 1 when it's actually
+            # 2. And for the moment I have no idea why.
+            if s.length > 1:
+                continue
+            if not isinstance(s, core.Sprixel):
+                raise base.PglInvalidTypeException(
+                    "GridSelector: the choices must be strings or Sprixels."
+                )
+            self.__cache.append(s)
+        self.__items_per_page = int(self.maximum_height / 2 * self.maximum_width / 2)
 
     @property
-    def color(self):
+    def maximum_width(self) -> int:  # #overwriting so build_cache()
+        #                              #can be called on size changes
         """
-        Get / set the edited color.
+        This property get/set the maximum width of the widget. This property is used
+        when changing the size constraints and the width property.
+        """
+        return self._Widget__maximum_width
 
-        The setter automatically forward the individual red, green and blue values to
-        to the proper properties of that widget.
+    @maximum_width.setter
+    def maximum_width(self, data: int) -> None:
```

Comment:
How exactly does GridSelector receive signals from, well effectively itself? I believe the PglBaseObject has been configured to prevent objects sending notifications to themselves.

Reply:
you are absolutely right! Damn...

Ok what do you think of these alternatives (it might be completely stupid): 

1. I could add a handle_private_notification() method to `PglBaseObject` that only allow processing of notifications from an object to itself (this way I'm not butchering the system's performances).
2. Alternatively we could have some sort of `post_processing(attribute)` method that is called directly by `Widget`. In `Widge`t the method would only `pass`, but deriving classes could implement the code needed for the post processing. In our case, in case `attribute` is *_width or height, we build the cache.

I'm trying to think of something to avoid the direct reference of "private" attributes as it as a tendency to break with inheritance (and these classes are designed to be inherited from).

What do you think?

Reply:
I'm probably in favour of the the `post_processing(attribute)` method, because it just feels a bit more elegant to me. But the `handle_private_notification()` method seems to me to be more versatile and will probably be useful far more often, even if I think its a bit uglier from an aesthetic perspective.

Reply:
I agree with you, both are suboptimal but unless you feel super courageous and implement a complete Qt style signal/slot to the whole library (:sweat_smile: ), I propose that we stick to the post processing method.

We can do it either way:

1. You make sure that the current code pass the CI validation, I merge the PR as is, and take care of the modifications.
2. If you feel like it, you can implement the changes in `Widget` and `GridSelector`.

In any case the work that needs to be done is:

1. Add a `post_processing(attribute: Optional[str] = None) -> None` method to `Widget`. The body should be `pass`.
2. Add a call to that method in all attribute setters of the `Widget` class. Update the documentation accordingly to make sure that it is clear that it is the default mechanism to have post processing actions on attributes (either `Widget`'s attribute or inheritors' ones.
3. Rename/refactor the `GridSelector.handle_notification` that you implemented because of my faulty recommendation to `post_processing(attribute)`. The body can probably remain the same.

I think that's it. Feel free to add stuff if I forgot something.